datalad/core/distributed/push.py

```diff
@@ -0,0 +1,726 @@
+    if cur_ds:
+        yield (cur_ds, ds_res)
+
+
+def _datasets_no_since_(dataset, paths, recursive, recursion_limit):
+    """Generator"""
+    # TODO can we do better an cheaper?
+    from datalad.consts import PRE_INIT_COMMIT_SHA
+    # use the diff "since before time"
+    return _datasets_since_(
+        dataset, PRE_INIT_COMMIT_SHA, paths, recursive, recursion_limit)
+
+
+def _push(dspath, content, target, force, jobs, res_kwargs,
+          done_fetch=None):
+    if not done_fetch:
+        done_fetch = set()
+    # nothing recursive in here, we only need a repo to work with
+    ds = Dataset(dspath)
+    repo = ds.repo
+
+    res_kwargs.update(type='dataset', path=dspath)
+
+    if not target:
+        try:
+            # let Git figure out what needs doing
+            wannabe_gitpush = repo.push(remote=None, git_options=['--dry-run'])
+            # we did not get an explicit push target, get it from Git
+            target = set(p.get('remote', None) for p in wannabe_gitpush)
+            # handle case where a pushinfo record did not have a 'remote'
+            # property -- should not happen, but be robust
+            target.discard(None)
+        except Exception as e:
+            lgr.debug(
+                'Dry-run push to determine default push target failed, '
+                'assume no configuration: %s', e)
+            target = set()
+        if not len(target):
+            yield dict(
+                res_kwargs,
+                status='error',
+                message='No push target given, and none could be '
+                        'auto-detected',
+            )
+            return
+        elif len(target) > 1:
+            # dunno if this can ever happen, but if it does, report
+            # nicely
+            yield dict(
+                res_kwargs,
+                status='error',
+                message=(
+                    'No push target given, '
+                    'multiple candidates auto-detected: %s',
+                    list(target),
+                )
+            )
+            return
+        else:
+            # can only be a single one at this point
+            target = target.pop()
+
+    if target not in repo.get_remotes():
+        yield dict(
+            res_kwargs,
+            status='error',
+            message=(
+                "Dataset %s does not know of a sibling '%s' to push to.",
+                ds, target))
+        return
+
+    lgr.debug("Attempt to push to '%s'", target)
+
+    # define config var name for potential publication dependencies
+    depvar = 'remote.{}.datalad-publish-depends'.format(target)
+    # list of remotes that are publication dependencies for the
+    # target remote
+    publish_depends = ensure_list(ds.config.get(depvar, []))
+    if publish_depends:
+        lgr.debug("Discovered publication dependencies for '%s': %s'",
+                  target, publish_depends)
+
+    # anything that follows will not change the repo type, cache.
```

Comment:
This is a confusing comment given that it doesn't look like anything above would have modified the local state.  Stale comment originating from `publish`?

Reply:
Yes, edited. Thx for catching that.